contracts/src/randomness/Randomness.cairo

```diff
@@ -0,0 +1,221 @@
+    requestor_address: felt,
+    request_id: felt,
+    minimum_block_number: felt,
+    callback_gas_limit: felt,
+    num_words: felt,
+) {
+}
+
+@event
+func Randomness__status_change(requestor_address: felt, request_id: felt, status: felt) {
+}
+
+//
+// Constructor
+//
+
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(admin_address) {
+    Admin.initialize_admin_address(admin_address);
+
+    return ();
+}
+
+//
+// External
+//
+
+// until cairo 1.0, we can not calculate onchain the amount of gas an invocation uses so we need to fail it ourselves
+@external
+func update_status{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    requestor_address, request_id, status
+) {
+    Admin.only_admin();
+    Randomness__request_status.write(requestor_address, request_id, status);
+    Randomness__status_change.emit(requestor_address, request_id, status);
+
+    return ();
+}
+
+@external
+func request_random{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    seed, callback_gas_limit, callback_address, publish_delay, num_words
+) -> (request_id: felt) {
+    alloc_locals;
+
+    let (caller_address) = get_caller_address();
+    let (current_block) = get_block_number();
+    let (request_id) = Randomness__request_id.read(caller_address);
+
+    let (hash_) = hash_request(
+        request_id, caller_address, current_block + publish_delay, callback_gas_limit, num_words
+    );
+
+    // hash request
+    Randomness__request_hash.write(caller_address, request_id, hash_);
+    Randomness__request.emit(
+        caller_address, request_id, current_block + publish_delay, callback_gas_limit, num_words
+    );
+    Randomness__request_status.write(caller_address, request_id, RequestStatus.RECEIVED);
+    Randomness__request_id.write(caller_address, request_id + 1);
+
+    return (request_id=request_id);
+}
+
+@external
+func cancel_random_request{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    requestor_address, request_id, minimum_block_number, callback_gas_limit, num_words
+) {
+    alloc_locals;
+
+    let (caller_address) = get_caller_address();
+    let (_hashed_value) = hash_request(
+        request_id, requestor_address, minimum_block_number, callback_gas_limit, num_words
+    );
+    let (stored_hash_) = Randomness__request_hash.read(caller_address, request_id);
+
+    with_attr error_message("INVALID REQUEST OWNER") {
+        assert _hashed_value = stored_hash_;
+        assert requestor_address = caller_address;
+    }
+    Randomness__status_change.emit(requestor_address, request_id, RequestStatus.CANCELLED);
+    return ();
+}
+
+@external
+func submit_random{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    request_id,
+    requestor_address,
+    minimum_block_number,
+    callback_gas_limit,
+    random_words_len,
+    random_words: felt*,
+) {
+    alloc_locals;
+
+    Admin.only_admin();
```

Comment:
Add comment: Will be replaced by proof in a few weeks once the VRF implementation is completed.